include/libsemigroups/schreier-sims-helper.hpp

```diff
@@ -0,0 +1,182 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+// This file contains an implementation of a backtrack search for computing
+// the intersection of two permutations given by Schreier-Sims algorithm.
+
+// TODO:
+// 1. Implement orbit refinement heuristic for intersection.
+// 2. Make the Screier-Sims object during runtime, since we compute the
+//    stabilizers of the intersection already.
+// 3. Refactor for more generality.
+
+#ifndef LIBSEMIGROUPS_SCHREIER_SIMS_HELPER_HPP_
+#define LIBSEMIGROUPS_SCHREIER_SIMS_HELPER_HPP_
+
+#include <cstddef>  // for size_t
+
+#include "containers.hpp"     // for Array2
+#include "schreier-sims.hpp"  // for SchreierSims, SchreierSims<>::ele...
+#include "transf.hpp"         // for LeastPerm
+#include "types.hpp"          // for SmallestInteger
+
+#include <iostream>
+
+namespace libsemigroups {
+
+  namespace schreier_sims_helper {
+
+    template <size_t N>
+    using point_type = typename SchreierSims<N>::point_type;
+
+    template <size_t N>
+    using element_type = typename SchreierSims<N>::element_type;
+
+    // This might not be correct for general traits
+    template <size_t N>
+    using One = typename SchreierSims<N>::One;
+
+    // This might not be correct for general traits
+    template <size_t N>
+    using Product = typename SchreierSims<N>::Product;
+
+    //! Find the intersection of two permutation groups.
+    //!
+    //! Modifies the first parameter \p T to be the Schreier-Sims object
+    //! corresponding to the intersection of \p S1 and \p S2.
+    //!
+    //! \tparam N the largest point not fixed by the permutations in the
+    //! permutation groups.
+    //!
+    //! \param T the Screier-Sims object holding the result.
+    //! \param S1 the first semigroup of the intersection.
+    //! \param S2 the second group of the intersection.
+    //!
+    // TODO(later) example
+    template <size_t N>
+    void intersection(SchreierSims<N>& T,
+                      SchreierSims<N>& S1,
+                      SchreierSims<N>& S2) {
+      S1.run();
+      S2.run();
+      if (S2.base_size() < S1.base_size()) {
+        intersection(T, S2, S1);
+        return;
+      }
+
+      // If N <= 1 then both S1, S2 are trivial.
+      if (N <= 1) {
+        T.run();
+        return;
+      }
+
+      // Note that if N-1 points are fixed then the N-th point is also fixed.
+      // So if base contains all N points, then we lose nothing by discarding
+      // the last point in the base.
+      size_t base_size = S1.base_size();
+      if (base_size == N)
+        base_size = N - 1;
+
+      SchreierSims<N> S2B;
```

Comment:
If `N` is big enough, then this will overflow the stack. Maybe `S2B` should be allocated on the heap, `auto S2B = new SchreierSims<N>();` or even better use a `unique_ptr` so you don't have to remember to delete this later. Alternatively, you might consider having two variants of this, one where S2B is passed as a parameter, and another where it is not (i.e. like the current version but where you pass it your newly allocated object).

Reply:
I did the heap allocation with the [`make_unique`](https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique) helper function.